src/vos/tests/evt_ctl.c

```diff
@@ -1682,6 +1683,220 @@ test_evt_various_data_size_internal(void **state)
 	}
 }
 
+static void
+test_evt_node_size_internal(void **state)
+{
+	struct test_arg		*arg = *state;
+	daos_handle_t		 toh;
+	int			order_size, rc;
+	bool			evt_created;
+
+	for (order_size = 0; order_size < ORDER_TEST_SIZE; order_size++) {
+		evt_created = false;
+		rc = evt_create(arg->ta_root, ts_feats, order_size,
+				arg->ta_uma, &ts_evt_desc_cbs, &toh);
+		if ((order_size >= 4) && (order_size <= 128)) {
+			assert_int_equal(rc, 0);
+			evt_created = true;
+		} else {
+			assert_int_not_equal(rc, 0);
+		}
+		if (evt_created) {
+			rc = evt_destroy(toh);
+			assert_int_equal(rc, 0);
+		}
+	}
+}
+
+/* Args : Input : string_val pointer
+*                 num: number of time to repeat string val
+*         Output: ret_string: Returns concatenated string.
+*/
+void repeat_string(char *ret_string, char *string_val, int num)
+{
+	int	counter;
+
+	counter = num;
+	strcpy(ret_string, string_val);
+	counter--;
+	while (counter-- > 0)
+		strcat(ret_string, string_val);
+}
+
+static void
+test_evt_overlap_split_internal(void **state)
+{
+	struct test_arg		*arg = *state;
+	daos_handle_t		toh;
+	daos_handle_t		ih;
+	struct evt_entry	ent;
+	int			rc;
+	struct evt_entry_in	entry = {0};
+	int			epoch;
+	int			record_size, data_size;
+	char			*data;
+	char			*expected_data;
+	char			*actual_data;
+	char			*read_data;
+	uint32_t		 inob;
+	int			data_len = 9;
+	int			tree_depth_fail = 0;
+	int			mem_cmp_fail = 0;
+	int			read_counter, counter, temp_val;
+	int			last_offset = 0;
+
+	rc = evt_create(arg->ta_root, ts_feats, ORDER_DEF_INTERNAL,
+				arg->ta_uma, &ts_evt_desc_cbs, &toh);
+	assert_int_equal(rc, 0);
+	/* Overall data : September
+	* overlap value1: Wonderful
+	* overlap value2: Rainydays
+	*/
+	record_size = ((data_len * NUM_EPOCHS) + 1);
+	D_ALLOC_ARRAY(data, record_size);
+	assert_non_null(data);
+	D_ALLOC_ARRAY(expected_data, record_size + data_len);
+	assert_non_null(expected_data);
+	repeat_string(data, "September", NUM_EPOCHS);
+	data_size = record_size;
+
+	D_ALLOC_ARRAY(expected_data, NUM_EPOCHS * record_size);
+	D_ALLOC_ARRAY(actual_data, NUM_EPOCHS * record_size);
+
+	for (epoch = 1; epoch < NUM_EPOCHS; epoch++) {
+	/* Write a big extent at epoch 1.
+	* Then, write some data in the front, end, middle
+	* at different epochs and see whether overlapped data
+	* is visible.
+	*/
+		if (epoch > 1) {
+			/* All value array data size is same.
+			*/
+			data_size = 9;
+			D_ALLOC_ARRAY(data, data_size);
```

Comment:
Use D_REALLOC_ARRAY ?  You allocate data above so there is a leak here

Reply:
thanks. Done.

Reply:
I see, any reason not to just allocate max(record_size, 9) above and just memset here?

Reply:
evt_insert uses data buffer (temporary location). This is the reason for different data_sizes. If we create entire data buffer, then we will do a single evt_insert. Is it not? The purpose here is to perform several evt_inserts and compare the expected/actual results [overlap values]. expected_data buffer is updated in parallel using temporary (data) buffer.
epoch 1 data : SeptemberSeptemberSeptemberSeptember... so on.
epoch 2 data : Wonderful
epoch 3 data : .................................Rainydays.............................
epoch 4 data :  .......................................................................................deadbeefs.........
epoch 5 and above :                .........................................................................................Wildfire..so on.
I didn't want to have same data on the whole range. That's why avoided memset.

Reply:
evt_insert uses the bio_addr allocated in bio_alloc_init on line 1845.   data buffer is just passed in as input but so is the size of what data buffer contains so the actual size can be larger.

Having different data sizes is fine.  It's unclear why you can't have a single allocation of data that is large enough for what you place into it in every iteration.

For instance, data could be a buffer of size 10000000 (size just has to be big enough).   In each iteration, you memset(data, 0, data_size) and use it as desired.   Make sense?

Reply:
Ok. Back from vacation. Understood. Will do the change tomorrow and send for re-review. Thanks for the input.